pyomo/core/expr/differentiate/diff_with_sympy.py

```diff
@@ -0,0 +1,299 @@
+    #
+    for i, target in enumerate(wrt_list):
+        if ans[i] is not None:
+            continue
+        part = sympy_expr
+        for j, wrt_var in enumerate(target):
+            if j == last_partial_idx:
+                part = sympy.diff(part, wrt_var)
+            else:
+                partial_target = target[:j+1]
+                if partial_target in partial_derivs:
+                    part = partial_derivs[partial_target]
+                else:
+                    part = sympy.diff(part, wrt_var)
+                    partial_derivs[partial_target] = part
+        ans[i] = sympy2pyomo_expression(part, objectMap)
+    #
+    # Return the answer
+    #
+    return ans if wrt is None else ans[0]
+
+
+# =====================================================
+# sympify_expression
+# =====================================================
+
+class Pyomo2SympyVisitor(EXPR.StreamBasedExpressionVisitor):
+
+    def __init__(self, object_map):
+        super(Pyomo2SympyVisitor, self).__init__()
+        self.object_map = object_map
+
+    def exitNode(self, node, values):
+        if node.__class__ is EXPR.UnaryFunctionExpression:
+            return _functionMap[node._name](values[0])
+        _op = _pyomo_operator_map.get(node.__class__, None)
+        if _op is None:
+            return node._apply_operation(values)
+        else:
+            return _op(*tuple(values))
+
+    def beforeChild(self, node, child):
+        #
+        # Don't replace native or sympy types
+        #
+        if type(child) in native_types:
+            return False, child
+        #
+        # We will descend into all expressions...
+        #
+        if child.is_expression_type():
+            return True, None
+        #
+        # Replace pyomo variables with sympy variables
+        #
+        if child.is_potentially_variable():
+            return False, self.object_map.getSympySymbol(child)
+        #
+        # Everything else is a constant...
+        #
+        return False, value(child)
+
+class Sympy2PyomoVisitor(EXPR.StreamBasedExpressionVisitor):
+
+    def __init__(self, object_map):
+        super(Sympy2PyomoVisitor, self).__init__()
+        self.object_map = object_map
+
+    def enterNode(self, node):
+        return (node._args, [])
+
+    def exitNode(self, node, values):
+        """ Visit nodes that have been expanded """
+        _sympyOp = node
+        _op = _operatorMap.get( type(_sympyOp), None )
+        if _op is None:
+            raise DeveloperError(
+                "sympy expression type '%s' not found in the operator "
+                "map" % type(_sympyOp) )
+        return _op(*tuple(values))
+
+    def beforeChild(self, node, child):
+        if not child._args:
+            item = self.object_map.getPyomoSymbol(child, None)
+            if item is None:
+                item = float(child.evalf())
+            return False, item
+        return True, None
+
+def sympify_expression(expr):
```

Comment:
Not a fault of this PR, but I think I should have called this "sympyify_expression" or "pyomo2sympy".